### PR TITLE
Release notes for 1.27.0

### PR DESCRIPTION
**Pull request by ikhoon**
image

### CODE REVIEW

File: site/src/pages/release-notes/1.27.0.mdx [hunk 1]
```diff
@@ -0,0 +1,228 @@
+---
+date: 2024-01-28
+---
+
+## 🌟 New features
+
+- **xDS-based EndpointGroup**: You can utilize xDS protocols to create <type://XdsEndpointGroup> for service
```
Comment (minwoox) [inline site/src/pages/release-notes/1.27.0.mdx:7]: How about `xDS support`?
How about also adding a link to it? https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol

File: site/src/pages/release-notes/1.27.0.mdx [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+  ```java
+  GrpcClients
+    .builder("https://my-grpc-server.com") 
+    .exceptionHandler((ctx, cause, metadata) -> {
+      // Handle the exception.
+      return Status.INTERNAL.withDescription("Something went wrong");
+    })
+  ```
+- **Thrift Service Decorators**: You can now utilize <type://@Decorator> annotations to conveniently apply
+  custom logic to Thrift services. #4230 #5356
```
Comment (minwoox) [inline site/src/pages/release-notes/1.27.0.mdx:54]: custom decorators?

File: site/src/pages/release-notes/1.27.0.mdx [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+    @Override
+    public void hello(String name, AsyncMethodCallback<String> handler) throws TException {
+      ...
+    }
+  }
+  ```
+- **Kubernetes Client Integration**: You can now use the popular [Fabric Kubernetes client](https://github.com/fabric8io/kubernetes-client)
+  seamlessly on top of Armeria when using the `armeria-kubernetes` module. #5167
+- **WebSocket Service Fallback**: Now, it is possible to set up a fallback service for 
+- <type://WebSocketService> to handle non-WebSocket requests seamlessly with
```
Comment (minwoox) [inline site/src/pages/release-notes/1.27.0.mdx:68]:
```suggestion
  <type://WebSocketService> to handle non-WebSocket requests seamlessly with
```

File: site/src/pages/release-notes/1.27.0.mdx [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+      // is not a WebSocket request.
+      .fallbackService(httpService)
+      .build();
+  Server
+    .builder()
+    .service("/ws-or-http", webSocketService)
+    .build();
+  ```
+- **WebSocket Frame Aggregation**: You can now enable to aggregate WebSocket continuation frames for simplified 
+- message processing using <type://WebSocketServiceBuilder#aggregateContinuation(boolean)>. #5357
```
Comment (minwoox) [inline site/src/pages/release-notes/1.27.0.mdx:83]:
```suggestion
- **WebSocket Frame Aggregation**: You can now enable aggregating WebSocket continuation frames for simplified 
  message processing using <type://WebSocketServiceBuilder#aggregateContinuation(boolean)>. #5357
```

File: site/src/pages/release-notes/1.27.0.mdx [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+      .builder()
+      .connectionPoolListener(ConnectionPoolListener.metricCollecting(meterRegistry))
+      .build();
+  WebClient
+    .builder("https://my-backend.com")
+    .factory(clientFactory)
+    .build();
+  ```
+- **Graceful Executor Shutdown**: You can now gracefully terminate a custom `ExecutorService` in sync with the
+- <type://Server>'s shutdown procedure using <type://ServerListenerBuilder#shutdownWhenStopped(ExecutorService,Duration)>
```
Comment (minwoox) [inline site/src/pages/release-notes/1.27.0.mdx:106]:
```suggestion
  <type://Server>'s shutdown procedure using <type://ServerListenerBuilder#shutdownWhenStopped(ExecutorService,Duration)>
```

Comment (minwoox) [reply site/src/pages/release-notes/1.27.0.mdx:106]: Do you mean whenStopping?

File: site/src/pages/release-notes/1.27.0.mdx [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+  ServerListener
+    .builder()
+    // Shutdown the executor service when the server is stopping.
+    .shutdownWhenStopping(preStopExecutorService)
+    // Shutdown the executor service when the server is stopped.
+    .shutdownWhenStopped(postStopExecutorService)
+    .build()
+  ```
+- **Session Protocol in RoutingContext**: You can now acquire <type://SessionProtocol> directly from
+  <type://RoutingContext#sessionProtocol()> for enhanced routing and <type://EchangeType> decisions.
```
Comment (minwoox) [inline site/src/pages/release-notes/1.27.0.mdx:118]:
```suggestion
  <type://RoutingContext#sessionProtocol()> for enhanced routing and <type://ExchangeType> decisions.
```

File: site/src/pages/release-notes/1.27.0.mdx [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+  EurekaUpdatingListener
+    .builder()
+    // Override the default health check path provided by HealthCheckService.
+    .healthCheckUrlPath("/healthcheck")
+    .build();
+  ```
+
+## 📈 Improvements
+
+- Annotated services now support <type://HttpResult> wrapped in RxJava objects #5380 #5386
```
Comment (minwoox) [inline site/src/pages/release-notes/1.27.0.mdx:133]:
```suggestion
- Annotated services now support <type://HttpResult> wrapped in RxJava objects. #5380 #5386
```

File: site/src/pages/release-notes/1.27.0.mdx [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+  [Spring Boot nested JAR URL](https://docs.spring.io/spring-boot/docs/current/reference/html/executable-jar.html#appendix.executable-jar.jarfile-class.compatibility). #5390 #5391
+- <type://WebsocketService> is restructured as an interface, allowing for custom implementations. #5368
+- Introduce <type://PreTlsHandshakeException> to better report exceptions occurring before a TLS handshake
+  initiates. #5344
+- Error messages for gRPC client stub generation have been clarified. #5387
+
+## 🛠️ Bug fixes
+
+- <type://ClientRequestContext#whenResponseCancelling()> and <type://ClientRequestContext#whenResponseCancelled()>
+  is now correctly completed even if the  `CancellationScheduler` has not been initialized #5212
```
Comment (minwoox) [inline site/src/pages/release-notes/1.27.0.mdx:144]:
```suggestion
  are now correctly completed even if the <type://CancellationScheduler> has not been initialized. #5212
```

File: site/src/pages/release-notes/1.27.0.mdx [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+- Fixed a leak within a WebSocket client when an invalid response is received. #5412
+- `json_name` attribute in protobuf files is now correctly used for matching query parameters during
+  gRPC HTTP/JSON transcoding. #5193 #5366
+- Implicit request objects are now correctly converted. #5389 #5393
+- Fixed an issue with handling an empty reason phase in WebSocket Close frames. #5388
+- <type://Flags> logging is now consistent, irrespective of <type://RequestContextExportingAppender> 
+  configurations in `logback.xml`. #5327 #5361
+- You no longer see <type://EndpointSelectionTimeoutException> when <type://HealthCheckedEndpointGroup> is 
+  initialized with duplicate `Endpoint`s. #5343
```
Comment (minwoox) [inline site/src/pages/release-notes/1.27.0.mdx:153]:
```suggestion
  initialized with duplicate <typeplural://Endpoint>. #5343
```

File: site/src/pages/release-notes/1.27.0.mdx [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+- <type://HttpHeaders#containsBoolean(CharSequence,boolean)> now checks across all header entries with the same
+  name. #5319
+- <type://GrpcService> provides more consistent `ServerCall.Listener` behavior under error conditions #5273
+
+## 📃 Documentation
+
+- Updated [Production Checklist](/docs/advanced-production-checklist) with guidance on 
+  using <type://ClientFactoryBuilder#maxNumEventLoopsPerEndpoint(int)>. This addition will help you mitigate 
+- potential client-side bottlenecks. #5297
```
Comment (minwoox) [inline site/src/pages/release-notes/1.27.0.mdx:162]:
```suggestion
  potential client-side bottlenecks. #5297
```

File: site/src/pages/release-notes/1.27.0.mdx [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+
+## 🛠️ Bug fixes
+
+- <type://ClientRequestContext#whenResponseCancelling()> and <type://ClientRequestContext#whenResponseCancelled()>
+  is now correctly completed even if the  `CancellationScheduler` has not been initialized #5212
+- Fixed a leak within a WebSocket client when an invalid response is received. #5412
+- `json_name` attribute in protobuf files is now correctly used for matching query parameters during
+  gRPC HTTP/JSON transcoding. #5193 #5366
+- Implicit request objects are now correctly converted. #5389 #5393
+- Fixed an issue with handling an empty reason phase in WebSocket Close frames. #5388
```
Comment (minwoox) [inline site/src/pages/release-notes/1.27.0.mdx:149]:
```suggestion
- WebSocket Close frames with empty reason phrases are not correctly handled. #5388
```

File: site/src/pages/release-notes/1.27.0.mdx [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+- <type://Flags> logging is now consistent, irrespective of <type://RequestContextExportingAppender>
```
Comment (minwoox) [inline site/src/pages/release-notes/1.27.0.mdx:150]: Is this generated by the ChatGPT? 😆
I'm not sure what this means. Could you use the original sentence in the commit description?

**Review by minwoox — COMMENTED** [review]
Thanks a lot! 👍

File: site/src/pages/release-notes/1.27.0.mdx (hunk 11 above)
Comment (ikhoon) [reply site/src/pages/release-notes/1.27.0.mdx:149]: Maybe `not` -> `now`?

File: site/src/pages/release-notes/1.27.0.mdx (hunk 5 above)
Comment (ikhoon) [reply site/src/pages/release-notes/1.27.0.mdx:106]: `shutdownWhenStopping` and `shutdownWhenStopped` is the actual API names.
Do you think we need to rename them before releasing 1.27.0?
https://github.com/line/armeria/blob/1862b7156c3606fcab7f2c09fa46f49588aae028/core/src/main/java/com/linecorp/armeria/server/ServerListenerBuilder.java#L327-L350

File: site/src/pages/release-notes/1.27.0.mdx (hunk 1 above)
Comment (ikhoon) [reply site/src/pages/release-notes/1.27.0.mdx:7]: > How about xDS support?

- We don't fully support xDS protocol so the scope is limited to `EndpointGroup`.

> How about also adding a link to it? 

I'm late. I was about to add it.

File: site/src/pages/release-notes/1.27.0.mdx (hunk 8 above)
Comment (ikhoon) [reply site/src/pages/release-notes/1.27.0.mdx]: `CancellationScheduler` is an internal API so the sentence has been rewritten with different words.

File: site/src/pages/release-notes/1.27.0.mdx (hunk 12 above)
Comment (ikhoon) [reply site/src/pages/release-notes/1.27.0.mdx]: Right... 😆 Many of them were abandoned while polishing the notes.

File: site/src/pages/release-notes/1.27.0.mdx (hunk 5 above)
Comment (minwoox) [reply site/src/pages/release-notes/1.27.0.mdx]: What I meant was that you specified the `shutdownWhenStopped` twice in the original note. 😆 
The sentence is rewritten so it's okay now. 😉

File: site/src/pages/release-notes/1.27.0.mdx [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+  now correctly respected for matching query parameters during gRPC HTTP/JSON transcoding. #5193 #5366
+- Implicit request objects are now correctly converted. #5389 #5393
+- <type://CloseWebSocketFrame> with empty reason phrases are now correctly handled. #5388
+- <type://Flags> now correctly logs all messages when <type://RequestContextExportingAppender> is configured in 
+  `logback.xml`. #5327 #5361
+- You no longer see <type://EndpointSelectionTimeoutException> when <type://HealthCheckedEndpointGroup> is 
+  initialized with duplicate <typeplural://Endpoint>. #5343
+- <type://HttpHeaders#containsBoolean(CharSequence,boolean)> now checks all header entries with the same
+  name. #5319
+- <type://GrpcService> provides more consistent `ServerCall.Listener` behavior under error conditions #5273
```
Comment (minwoox) [inline site/src/pages/release-notes/1.27.0.mdx:169]:
```suggestion
- <type://GrpcService> provides more consistent `ServerCall.Listener` behavior under error conditions. #5273
```

**Review by minwoox — APPROVED** [review]
Thanks a lot, @ikhoon 👍 👍 👍

**Review by jrhee17 — APPROVED** [review]
Looks nice! Thanks @ikhoon 🙇 👍 🙇

**Comment by ikhoon** [comment]
Thanks for the reviews! 🙇‍♂️

**Pull request merged**